icicle-cpu/src/pe.rs

```diff
@@ -0,0 +1,205 @@
+                            let old_val = u16::from_le_bytes(buf.try_into().unwrap());
+                            let new_val = old_val.wrapping_add((relocation_offset >> 16) as u16);
+                            cpu.mem.write_bytes(addr, &new_val.to_le_bytes(), perm::READ)
+                                .map_err(|e| format!("Failed to write relocation data: {e:?}"))?;
+                        },
+                        IMAGE_REL_BASED_LOW => {
+                            let mut buf = vec![0_u8; 2];
+                            cpu.mem.read_bytes(addr, &mut buf[..], perm::READ)
+                                .map_err(|e| format!("Failed to read relocation data: {e:?}"))?;                            
+                            let old_val = u16::from_le_bytes(buf.try_into().unwrap());
+                            let new_val = old_val.wrapping_add(relocation_offset as u16);
+                            cpu.mem.write_bytes(addr, &new_val.to_le_bytes(), perm::READ)
+                                .map_err(|e| format!("Failed to write relocation data: {e:?}"))?;
+                        },
+                        IMAGE_REL_BASED_HIGHLOW => {
+                            let mut buf = vec![0_u8; 4];
+                            cpu.mem.read_bytes(addr, &mut buf[..], perm::READ)
+                                .map_err(|e| format!("Failed to read relocation data: {e:?}"))?;                            
+                            let old_val = u32::from_le_bytes(buf.try_into().unwrap());
+                            let new_val = old_val.wrapping_add(relocation_offset as u32);
+                            cpu.mem.write_bytes(addr, &new_val.to_le_bytes(), perm::READ)
+                                .map_err(|e| format!("Failed to write relocation data: {e:?}"))?;                            
+                        },
+                        IMAGE_REL_BASED_DIR64 => {
+                            let mut buf = vec![0_u8; 8];
+                            cpu.mem.read_bytes(addr, &mut buf[..], perm::READ)
+                                .map_err(|e| format!("Failed to read relocation data: {e:?}"))?;                            
+                            let old_val = u64::from_le_bytes(buf.try_into().unwrap());
+                            let new_val = old_val.wrapping_add(relocation_offset);                            
+                            cpu.mem.write_bytes(addr, &new_val.to_le_bytes(), perm::READ)
+                                .map_err(|e| format!("Failed to write relocation data: {e:?}"))?;
+                        },
+                        _ => {
+                            return Err(String::from(format!("Relocation type {:#06x} not supported", reloc.typ)));
+                        }
+                    }
+                }
+            }
+        }
+        else {
+            return Err(String::from("PE file does not contain relocation data, unable to relocate file"));
```

Comment:
if the allocation base is different than the preferred base address a relocation section must be present otherwise the binary is not loaded. However, the relocation section can be empty. I'll address this possibility.